torch/csrc/api/include/torch/nn/parallel/data_parallel.h

```diff
@@ -0,0 +1,161 @@
+/// outputs of the individual calls are stored in a vector and returned.
+///
+/// Further remarks:
+/// 1. The length of the module container must match the length of the inputs.
+/// 2. If a list of devices is supplied, it must match the list of modules in
+/// length. Each device will be set to the current default device during the
+/// invocation of the respective module. This means any tensors allocated on the
+/// default device inside the module will be constructed on this device.
+template <typename ModuleType>
+std::vector<Tensor> parallel_apply(
+    std::vector<ModuleType>& modules,
+    const std::vector<Tensor>& inputs,
+    const at::optional<std::vector<Device>>& devices = at::nullopt) {
+  AT_CHECK(
+      modules.size() == inputs.size(), "Must have as many inputs as modules");
+  if (devices) {
+    AT_CHECK(
+        modules.size() == devices->size(),
+        "Must have as many devices as modules");
+  }
+
+  std::vector<Tensor> outputs(modules.size());
+  std::mutex mutex;
+
+  at::parallel_for(
+      /*begin=*/0,
+      /*end=*/modules.size(),
+      /*grain_size=*/1,
+      [&modules, &inputs, &devices, &outputs, &mutex](
+          int64_t index, int64_t stop) {
+        for (; index < stop; ++index) {
+          try {
+            torch::OptionsGuard options_guard(
+                devices ? (*devices)[index] : inputs[index].device());
+            auto output = modules[index]->forward(inputs[index]);
+            std::lock_guard<std::mutex> lock(mutex);
+            outputs[index] = output;
+          } catch (const std::exception& exception) {
+            std::cerr << "Replica #" << index
+                      << " terminated in parallel_apply() with exception: "
+                      << exception.what();
+            std::terminate();
+          }
+        }
+      });
+
+  return outputs;
+}
+
+/// Evaluates `module(input)` in parallel across the given `devices`. If
+/// `devices` is not supplied, the invocation is parallelized across all
+/// available CUDA devices. If `output_device` is supplied, the final, combined
+/// tensor will be placed on this device. If not, it defaults to the first
+/// device in `devices`.
+///
+/// In detail, this method performs the following four distinct steps:
+/// 1. *Scatter* the input to the given devices,
+/// 2. *Replicate* (deep clone) the model on each device,
+/// 3. *Evaluate* each module with its input on its device,
+/// 4. *Gather* the outputs of each replica into a single output tensor, located
+/// on the `output_device`.
+template <typename ModuleType>
+Tensor data_parallel(
+    ModuleType module,
+    Tensor input,
+    at::optional<std::vector<Device>> devices = at::nullopt,
+    at::optional<Device> output_device = at::nullopt,
+    int64_t dim = 0) {
+  if (!devices) {
+    const auto device_count = torch::cuda::device_count();
+    AT_CHECK(device_count > 0, "Expected at least one CUDA device");
+    devices.emplace();
+    devices->reserve(device_count);
```

Comment:
Why do you reserve only once you emplace? This will likely trigger a reallocation. Also, aren't you constructing a vector where the first device appears twice?

Reply:
`emplace()` only creates the vector, `reserve` then allocates its memory. I think there are no unnecessary allocations?

Reply:
Ok, I didn't notice it's an optional thing, so I understood that as `emplace_back()`.